packages/aws-cdk/lib/api/util/checks.ts

```diff
@@ -0,0 +1,88 @@
+import { debug } from '../../logging';
+import { ISDK } from '../aws-auth/sdk';
+
+export async function determineAllowCrossAccountAssetPublishing(sdk: ISDK, customStackName?: string): Promise<boolean> {
+  const stackName = customStackName || 'CDKToolkit';
+  if (!(await _stagingBucketInToolkitStack(sdk, stackName))) {
+    // indicates an intentional cross account setup
+    return true;
+  }
+
+  if (await _bootstrapVersion(sdk, stackName) >= 21) {
+    // bootstrap stack version 21 contains a fix that will prevent cross
+    // account publishing on the IAM level
+    // https://github.com/aws/aws-cdk/pull/30823
+    return true;
```

Comment:
```
// therefore, we can safely allow this in the CLI. It will either fail, or if it succeeds then
// the user has customized their bootstrap stack to explicitly allow it, which is also fine
```